vertical-pod-autoscaler/pkg/utils/vpa/capping.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	apiv1 "k8s.io/api/core/v1"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/poc.autoscaling.k8s.io/v1alpha1"
+)
+
+// GetCappedRecommendationForContainer returns a recommendation extracted for the given container, adjusted to obey policy and limits.
+func GetCappedRecommendationForContainer(
+	container apiv1.Container,
+	podRecommendation *vpa_types.RecommendedPodResources,
+	policy *vpa_types.PodResourcePolicy) (apiv1.ResourceList, error) {
+	containerRecommendation := getRecommendationForContainer(podRecommendation, container)
+	if containerRecommendation == nil {
+		return nil, fmt.Errorf("no recommendation available for container name %v", container.Name)
+	}
+	capRecommendationToContainerLimit(containerRecommendation, container)
```

Comment:
I think capping to the limit should be applied *after* the VPA policy. It normally shouldn't matter, but in the unlikely case of a conflict (currently: minAllowed > limit), capping to limit should take precedence. I guess the updater could set some error condition in such case (maybe a TODO for now?)